Display SPDX license expressions and links

kemitchell opened the pull request:
This pull request harmonizes display of license links in the sidebars of package pages with recent clarifications to the `package.json` metadata guidelines. It directly addresses #992, and may or may not obviate npm/normalize-license-data#1.

Under the hood, the presenter for packages is changed to utilize [`spdx-to-html`](https://github.com/kemitchell/spdx-to-html.js), a purpose-built package, in lieu of `normalize-license-data`. `spdx-to-html` uses [`spdx`](https://github.com/kemitchell/spdx.js), the same package `normalize-package-data` and `init-package-json` now use for validation, to generate an AST for license expressions and flatten them into strings of HTML.

If there is anything icky about this patch, it's that `spdx-to-html` generates HTML that's inserted without escaping by Handlebars.

If there's work still to be done, it's for kemitchell/spdx-to-html.js#1, to make sure there's no wonky or misleading output if any when a package author decides to use SPDX expressions to their fullest. If expressions are new, have a look at [the README to `spdx.js`](http://npmjs.com/packages/spdx), which includes examples with multiple conjunctions and parentheses for grouping. There are a few multi-licensed packages in the registry that are doing basic conjunctions (many with my help). So far, I'm not aware of any using an expression that embodies this kind of edge case.

The arm of licensing law is a long arm indeed:

- npm/npm#8179 — here the guidelines were justly and ably amended
- [2.10.0 release notes](https://github.com/npm/npm/releases/tag/v2.10.0) — here that wise decision was announced to the people
- npm/normalize-license-data#1 — here the topic of spdx.org URLs was first broached

Poor @othiym23's been with me at each and every turn. @SimenB first saw the need for wiser license links, and acted on it. [JSONStream](http://npmjs.com/packages/JSONStream), I think it was.

kemitchell commented:
I've fiddled a bit with the code. Mind giving it one more shot?

I also rebased onto current master. Still no merge conflicts.

kemitchell commented:
Thoughts are these: I am not good at this!

SimenB commented:
Just a quick note: I tried to `require` the hbs-template and pass the data from @kemitchell's test through it, and the license is included in the html output, so it seems to work like it should.
Other than having to remove the partial `form_security` from the template to get it to run cleanly from a test, this is the full output from the test: http://hastebin.com/apokivameq.html. License is included on line 40.

To test it, I added `require('handlebars')` at the top of the test file, as it uses `require.extensions`, then `require(__dirname + "/../../templates/package/show.hbs")({'package': pkg})` in the test itself.

SimenB commented:
@rockbot Any news? :smile:

aredridel commented:
Looks like this branch could be rebased against current master -- I could take a peek at it but there's also conflicts.

kemitchell commented:
I will have another look this weekend, when I finish refactoring the SPDX-related modules. I also need to handle the magic values we recently coined for nonstandard licenses.

My main concern is that I'm really sure why my first patch didn't work. I'm still not entirely sure I understand how the template logic is actually rendered. I'll have another look.

aredridel commented:
Alright! Ping for help if you want a look -- happy to guide where I can.

kemitchell commented:
I've updated the PR and related modules to address the recent changes to valid license metadata. `npm test` is green with local Redis. I believe this is as far as I got before.

@rockbot, any chance you might run this on npm infrastructure?

sam-github commented:
Any chance of getting this fixed?

Its pretty broken at the moment, "SEE LICENSE IN LICENSE license" shows up on the web site, linking to http://opensource.org/licenses/SEE%20LICENSE%20IN%20LICENSE

/cc @rmg @superkhau

kemitchell commented:
Rebased. Green and clean.

rmg commented:
:+1: this would be especially nice since the npm cli currently gives warnings if you use the only method the website currently supports for fixing this URLs in package.

superkhau commented:
+1

kemitchell commented on presenters/package.js:
```diff
@@ -22,9 +23,25 @@ module.exports = function(pkg) {
+    return Error('invalid pkg: ' + pkg.name);
+  }
+
+  if (!pkg.license || typeof pkg.license !== 'string') {
```
@rockbot, the old Ruby-esque object style is now deprecated: https://github.com/npm/npm/blob/master/doc/files/package.json.md#license

Dominic was kind enough to update to the new format: https://github.com/npm/npm/blob/master/doc/files/package.json.md#license

cc @othiym23

SimenB replied:
It should still be supported though, shouldn't it? npm has almost 200k packages (according to http://www.modulecounts.com/, no idea if it's actually accurate), I bet a huge chunk of them will not update to the new format, but they might use the old object or array-style.
Maybe add an annoying warning next to it, to encourage updating? But it shouldn't say `none license` (or just drop it entirely), or something similar when they actually define a license, albeit in a deprecated format.

kemitchell replied:
There's a call to be made here.

The current behavior more strongly encourages conformance to the new guidelines.

@SimenB mentions a warning, which is (as usual!) great input. That's probably closest to the approach CLI has taken.

I'm cloning npm.dathub.com now, and will run a quick survey on Array and Object license props.

kemitchell replied:
There is trouble in datland. I'll try to get it sorted. Stats on Array/Object license values may not be immediately forthcoming.

kemitchell replied:
Thanks, @rockbot! Really appreciate your time on this.

kemitchell replied:
Based on most recent npm.dathub.org, about 1,600 package have an Object for "license". Less than 200 have an Array.

Special thanks to @mafintosh for addressing maxogden/dat#408 so quickly.

kemitchell replied:
Code for the survey: https://github.com/kemitchell/dat-npm-licenses

kemitchell commented:
Just checking in. Thanks again to all involved---y'all're busy wombats, I know!

If there's anything I can do independently to help this long, please let me know.

aredridel commented:
We may have some time coming up, probably not this week but just a bit after I hope :)

kemitchell commented:
Thanks, Aria!

kemitchell commented:
Just following up. Don't mean at all to nag.

Any chance full testing on non-npm folks' machines has come along since I PR'd? Thanks!

kemitchell commented:
The helper package I wrote for this PR had an errant `main` property in `package.json`. I've fixed it, pushed a patch release, and amended here to pick up the newer version. I'm not sure if that had anything to do with the failure here, but it's fixed now!

aredridel commented:
It does not fix it, but if you wanted to rebase against current master (`git fetch origin; git rebase origin/master` on this branch) I bet the tests would start passing, since master is in a good state right now and the failures seem to have come from the snapshot you developed this from.

kemitchell commented:
@aredridel I have been running the test suite with a local Redis. Is there any way for me to dev-deploy on my own machine to reproduce the problem you're seeing?

kemitchell commented:
I've rebased and run the test suite locally. There's a regression in my commit in unrelated tests. I'll look into it.

kemitchell commented:
Fixed the regression. Dropped some code from the package presenter that must have ridden along with all the rebases over time. The code was:

```javascript
// presenters/package.js
if (pkg.versions && pkg.versions.indexOf(pkg.version) === -1) {
  return Error('invalid pkg: ' + pkg.name);
}
```

kemitchell commented:
Green!

kemitchell commented:
Rebased. The only conflict was a package.json version bump.
